roles/openshift_health_checker/openshift_checks/docker_image_availability.py

```diff
@@ -0,0 +1,161 @@
+
+    This check attempts to ensure that required docker images are
+    either present locally, or able to be pulled down from available
+    registries defined in a host machine.
+    """
+
+    name = "docker_image_availability"
+    tags = ["preflight"]
+
+    skopeo_image = "juanvallejo/skopeo:v3.4"
+
+    docker_image_base = {
+        "origin": {
+            "repo": "openshift",
+            "image": "origin",
+        },
+        "openshift-enterprise": {
+            "repo": "openshift3",
+            "image": "ose",
+        },
+    }
+
+    def run(self, tmp, task_vars):
+        required_images = self.required_images(task_vars)
+        missing_images = set(required_images) - set(self.local_images(required_images, task_vars))
+
+        # exit early if all images were found locally
+        if not missing_images:
+            return {"changed": False}
+
+        registries = self.known_docker_registries(task_vars)
+        available_images, changed = self.available_images(missing_images, registries, task_vars)
+        unavailable_images = set(missing_images) - set(available_images)
+
+        if unavailable_images:
+            return {
+                "failed": True,
+                "msg": (
+                    "One or more required images are not available: {}.\n"
+                    "Configured registries: {}"
+                ).format(", ".join(sorted(unavailable_images)), ", ".join(registries)),
+                "changed": changed,
+            }
+
+        return {"changed": changed}
+
+    def required_images(self, task_vars):
+        """ Return a list of all required tagged images """
+        deployment_type = get_var(task_vars, "deployment_type")
+        image_base_name = self.docker_image_base[deployment_type]
+
+        # openshift_release = get_var(task_vars, "openshift_release")
+        # openshift_image_tag = get_var(task_vars, "openshift_image_tag")
+        openshift_release = "v3.4"
+        openshift_image_tag = "3.4.1.7"
+
+        is_containerized = get_var(task_vars, "openshift", "common", "is_containerized")
+
+        if is_containerized:
+            docker_or_rpm_images = set(self.containerized_docker_images(image_base_name, openshift_release))
+        else:
+            docker_or_rpm_images = set(self.rpm_docker_images(image_base_name, openshift_release))
+
+        # append images with qualified image tags to our list of required images.
+        # these are images with a (v0.0.0.0) tag, rather than a standard release
+        # format tag (v0.0). We want to check this set in both containerized and
+        # non-containerized installations.
+        docker_or_rpm_images.update(self.qualified_docker_images(self.image_from_base_name(image_base_name),
+                                                                 "v" + openshift_image_tag))
+
+        return docker_or_rpm_images
+
+    def local_images(self, required_images, task_vars):
+        """ Return all found local docker images """
```

Comment:
Please self-review the code for this type of general comment, so that we don't need to comment on every single occurrence.